src/validation.cpp

```diff
@@ -812,6 +812,10 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
         return state.Invalid(TxValidationResult::TX_INPUTS_NOT_STANDARD, "bad-txns-nonstandard-inputs");
     }
 
+    if (DatacarrierBytes(tx, m_view) > m_pool.m_max_datacarrier_bytes) {
```

Comment:
```suggestion
    if (m_pool.m_max_datacarrier_bytes && DatacarrierBytes(tx, m_view) > *m_pool.m_max_datacarrier_bytes) {
```
This fixes the problem mentioned in https://github.com/bitcoin/bitcoin/pull/28408#issuecomment-1710076571 (which I can reproduce), and is necessary because `m_pool.m_max_datacarrier_bytes` is type `std::optional<unsigned int>`. If the configuration includes `datacarrier=0` this variable is `nullopt`. It's surprising to me that without this suggestion (comparing an `int32_t` to an unset `std::optional<unsigned int>` for greater-than), the test succeeds (which deems the transaction to be invalid, so the mempool stays completely empty).

This suggestion also requires changing the return type of `DatacarrierBytes()` from `int32_t` to `uint32_t` (I think an unsigned type makes more sense anyway).

Note that this change does not cause the mempool to accept data-carrying transactions (as it might appear) if `datacarrier=0` because they're filtered out by the call to `IsStandardTx()` earlier in this function.

Reply:
It worked all transactions are no longer purged at startup, however even with `datacarrier = 0`, inscriptions still appear in my mempool(I don't know if it still works for inscriptions of more than 80 bytes).

We can see it [here](https://orangepill.ovh/fr/block/00000000000000000002561e1e9e110add7e63f716c3b460499d7b1445f93d1f), as my node refuses the inscriptions they are supposed to appear in blue however we see that many of them have still passed.

Reply:
But a code correction is needed in order datacarriersize=0 to behave as expected. 

```suggestion
    if (DatacarrierBytes(tx, m_view) > *m_pool.m_max_datacarrier_bytes) {
```
[https://github.com/luke-jr/bitcoin/pull/10/commits/669cdb01e75c432501f096e752a0342caf08d4a0](url)

Reply:
`datacarriersize=0` works as expected with @LarryRuane's suggestions.

Reply:
Also with this #https://github.com/bitcoin/bitcoin/pull/28408#discussion_r1332912845 suggestion?

Reply:
I don't know, I haven't tested your suggestions.

Reply:
> m_pool.m_max_datacarrier_bytes &&

@LarryRuane: why is that condition necessary? It provides always a pointer different than NULL, right? It is always interpreted as “True” by the ‘If’ conditional expression.

Reply:
It's necessary because `m_pool.m_max_datacarrier_bytes` isn't a pointer, it's a `std::optional`. It may be unset (it will be if the config includes `-datacarrier=0`), so it must be tested in this way (or we could call `m_pool.m_max_datacarrier_bytes.has_value()`, which would have the same effect) before attempting to access the contained value (which is what the `*` does).

Reply:
@LarryRuane: You are right about: m_pool.m_max_datacarrier_bytes isn't a pointer, it's a std::optional.

Two user parameters are in play:
-**datacarrier** and
-**datacarriersize**

Following code extract determine the behavior: 

`    if (argsman.GetBoolArg("-datacarrier", DEFAULT_ACCEPT_DATACARRIER)) {
        mempool_opts.max_datacarrier_bytes = argsman.GetIntArg("-datacarriersize", MAX_OP_RETURN_RELAY);
    } else {
        mempool_opts.max_datacarrier_bytes = std::nullopt;
    }`

In case **-datacarrier=0**, then your proposal will **not** work as intended since max_datacarrier_bytes will be std::nullopt and the conditional:
`if (m_pool.m_max_datacarrier_bytes && DatacarrierBytes(tx, m_view) > *m_pool.m_max_datacarrier_bytes)`
is not fulfilled.

For that reason here a suggested code change:

```suggestion
    if (!m_pool.m_max_datacarrier_bytes || (DatacarrierBytes(tx, m_view) > *m_pool.m_max_datacarrier_bytes)) {
```

Reply:
No, if you do this (and the config includes `-datacarrier=0`), then _all_ transactions are rejected. (Try it, you'll see that `bitcoin-cli getmempoolinfo` shows size and bytes are zero.) We don't want that, we want all data-carrying transactions to be rejected, but non-data carrying should not be rejected.

Reply:
Yes. You are right again!

Reply:
What is the intended use case for -datacarriers=0?

It seems to be:
-datacarrier=1
-datacarriersize=0
If the new policy rule should reject any data carrying transaction, right?

Reply:
Maybe following code suggestion could consider the use cases:
> **no datacarrier** option given: default full node behavior
> **-datacarrier=0**: full node does not accept any transaction with data carrier
> **-datacarrier=1** and **-datacarriersize=0**: same behavior like use case above
> **-datacarrier=1** and **-datacarriersize="any number"**: full node accept data
   carrier transactions until size of data carrier being less than "any number"
> **no datacarrier** option given and **-datacarriersize="any number"**: same behavior like use case above

In validation.cpp:
```suggestion
    if (IsDataCarrier(tx, m_view, m_pool.m_max_datacarrier_bytes)) {
```

In policy.h:
`bool IsDataCarrier(const CTransaction& tx, const CCoinsViewCache& view, const std::optional<unsigned>& max_datacarrier_bytes);
`

In policy.cpp:
`bool IsDataCarrier(const CTransaction& tx, const CCoinsViewCache& view, const std::optional<unsigned>& max_datacarrier_bytes)
{
    bool fRet{0};
    uint32_t DataCarrierBytesSize{0};
    DataCarrierBytesSize = DatacarrierBytes(tx, view);
    if (DataCarrierBytesSize > 0){
        if (!max_datacarrier_bytes){
            fRet = true;
        }else{
            if (max_datacarrier_bytes.has_value() && (DataCarrierBytesSize > *max_datacarrier_bytes)){
                fRet = true;
            }
        }
    }
    return fRet;
}`

Reply:
```c++
    if (DatacarrierBytes(tx, m_view) > *m_pool.m_max_datacarrier_bytes) {
Commit suggestion￼Add suggestion to batch
```

This would be undefined behaviour if it's a nullopt. Using `.value_or(0)` instead.